base_layer/core/src/transaction_protocol_manager.rs

```diff
@@ -0,0 +1,1046 @@
+    }
+
+    /// The Sender will use this method to construct the public data message they will send to the Receiver
+    fn construct_sender_public_data(&self) -> Result<SenderPublicData, TransactionProtocolError> {
+        // Validate the current state to check we have the data we need
+        if self.state_data.amount.is_none() ||
+            self.state_data.lock_height.is_none() ||
+            self.state_data.fee.is_none() ||
+            self.state_data.offset.is_none() ||
+            self.state_data.sender_public_nonce.is_none() ||
+            self.state_data.sender_excess.is_none() ||
+            self.state_data.inputs.len() == 0
+        {
+            return Err(TransactionProtocolError::IncompleteStateError);
+        }
+
+        Ok(SenderPublicData {
+            fee: self.state_data.fee.unwrap(),
+            lock_height: self.state_data.lock_height.unwrap(),
+            amount: self.state_data.amount.unwrap(),
+            sender_excess: self.state_data.sender_excess.unwrap(),
+            sender_public_nonce: self.state_data.sender_public_nonce.unwrap(),
+            offset: self.state_data.offset.unwrap(),
+        })
+    }
+
+    fn set_output(
+        mut self,
+        output: TransactionOutput,
+        blinding_factor: BlindingFactor,
+    ) -> Result<TransactionProtocolState, TransactionProtocolError>
+    {
+        self.state_data.receiver_output = Some(output);
+        self.state_data.receiver_output_blinding_factor = Some(blinding_factor);
+        self.state_data.receiver_output_public_key = Some(PublicKey::from_secret_key(&blinding_factor));
+
+        Ok(TransactionProtocolState::ReceiverPartialSignature(ReceiverPartialSignature::new(self.state_data)))
+    }
+}
+
+/// In this state the Receiver is ready to provide the data required for them to construct their partial signature
+#[derive(Clone, Debug)]
+struct ReceiverPartialSignature {
+    state_data: TransactionProtocolStateData,
+}
+
+impl ReceiverPartialSignature {
+    fn new(previous_state_data: TransactionProtocolStateData) -> ReceiverPartialSignature {
+        ReceiverPartialSignature { state_data: previous_state_data }
+    }
+
+    fn set_private_nonce(mut self, nonce: SecretKey) -> Result<TransactionProtocolState, TransactionProtocolError> {
+        // Validate that all the required state is present.
+        if self.state_data.sender_public_nonce.is_none() ||
+            self.state_data.sender_excess.is_none() ||
+            self.state_data.receiver_output_public_key.is_none() ||
+            self.state_data.fee.is_none() ||
+            self.state_data.lock_height.is_none() ||
+            self.state_data.receiver_output_blinding_factor.is_none()
+        {
+            return Err(TransactionProtocolError::IncompleteStateError);
+        }
+
+        self.state_data.receiver_private_nonce = Some(nonce.clone());
+        self.state_data.receiver_public_nonce = Some(PublicKey::from_secret_key(&nonce));
```

Comment:
Swap the 2 lines above and you can drop the `clone`

Reply:
Yup, done